pkg/fieldpath/merge.go

```diff
@@ -0,0 +1,156 @@
+	}
+	return config
+}
+
+// ToPaved tries to convert a runtime.Object into a *Paved via
+// runtime.DefaultUnstructuredConverter if needed. Returns the paved if
+// the conversion is successful along with whether the
+// runtime.DefaultUnstructuredConverter has been employed during the
+// conversion.
+func ToPaved(o runtime.Object) (*Paved, bool, error) {
+	if u, ok := o.(interface{ UnstructuredContent() map[string]interface{} }); ok {
+		return Pave(u.UnstructuredContent()), false, nil
+	}
+
+	oMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(o)
+	if err != nil {
+		return nil, false, err
+	}
+	return Pave(oMap), true, nil
+}
+
+// PatchFieldValueToObject applies the value to the "to" object at the given
+// path with the given merge options, returning any errors as they occur.
+// If no merge options is supplied, then destination field is replaced
+// with the given value.
+func PatchFieldValueToObject(fieldPath string, value interface{}, to runtime.Object,
+	mergeOptions *MergeOptions) error {
+	paved, copied, err := ToPaved(to)
+	if err != nil {
+		return err
+	}
+	dst, err := paved.GetValue(fieldPath)
+	if IsNotFound(err) || mergeOptions == nil {
+		dst = nil
+	} else if err != nil {
+		return err
+	}
+
+	dst, err = merge(dst, value, mergeOptions)
+	if err != nil {
+		return err
+	}
+
+	if err := paved.SetValue(fieldPath, dst); err != nil {
+		return err
+	}
+
+	if copied {
+		return runtime.DefaultUnstructuredConverter.FromUnstructured(paved.object, to)
+	}
+	return nil
+}
+
+// MergePath merges the value at the given field path of the src object into
+// the dst object.
+func MergePath(fieldPath string, dst, src runtime.Object, mergeOptions *MergeOptions) error {
+	srcPaved, _, err := ToPaved(src)
+	if err != nil {
+		return err
+	}
+
+	val, err := srcPaved.GetValue(fieldPath)
+	// if src has no value at the specified path, then nothing to merge
+	if IsNotFound(err) || val == nil {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	return PatchFieldValueToObject(fieldPath, val, dst, mergeOptions)
+}
+
+// MergeReplace merges the value at fieldPath from desired into
+// a copy of current and then replaces the value at fieldPath of
+// desired with the merged value. current object is not modified.
+func MergeReplace(fieldPath string, current, desired runtime.Object, mergeOptions *MergeOptions) error {
+	copyCurrent := current.DeepCopyObject()
+	if err := MergePath(fieldPath, copyCurrent, desired, mergeOptions); err != nil {
+		return err
+	}
+	// replace desired object's value at fieldPath with
+	// the computed (merged) current value at the same path
+	return MergePath(fieldPath, desired, copyCurrent, nil)
+}
+
+// merges the given src onto the given dst.
+// If a nil merge options is supplied, the default behavior is MergeOptions'
+// default behavior. If dst or src is nil, src is returned
+// (i.e., dst replaced by src).
+func merge(dst, src interface{}, mergeOptions *MergeOptions) (interface{}, error) {
```

Comment:
Does this function require that `dst` and `src` are of the same underlying type?